{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Univariate correlation analysis\n",
    "\n",
    "TODO: describe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from pathlib import Path\n",
    "import warnings\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "import pancancer_evaluation.config as cfg\n",
    "import pancancer_evaluation.utilities.data_utilities as du\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = 'TP53'\n",
    "mad_threshold = 100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load expression data and mutation label data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading gene label data...\n",
      "Loading sample info...\n",
      "Loading pan-cancer data from cached pickle file...\n",
      "Loading gene expression data...\n",
      "Standardizing columns of expression data...\n"
     ]
    }
   ],
   "source": [
    "print('Loading gene label data...', file=sys.stderr)\n",
    "genes_df = du.load_top_50()\n",
    "sample_info_df = du.load_sample_info(verbose=True)\n",
    "\n",
    "# this returns a tuple of dataframes, unpack it below\n",
    "pancancer_data = du.load_pancancer_data(verbose=True)\n",
    "(sample_freeze_df,\n",
    " mutation_df,\n",
    " copy_loss_df,\n",
    " copy_gain_df,\n",
    " mut_burden_df) = pancancer_data\n",
    "\n",
    "rnaseq_df = du.load_expression_data(verbose=True)\n",
    "\n",
    "# standardize columns of expression dataframe\n",
    "print('Standardizing columns of expression data...', file=sys.stderr)\n",
    "rnaseq_df[rnaseq_df.columns] = StandardScaler().fit_transform(rnaseq_df[rnaseq_df.columns])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11060, 16148)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>10</th>\n",
       "      <th>100</th>\n",
       "      <th>1000</th>\n",
       "      <th>10000</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0047-01</th>\n",
       "      <td>-0.144100</td>\n",
       "      <td>-0.136450</td>\n",
       "      <td>-0.207065</td>\n",
       "      <td>1.049402</td>\n",
       "      <td>0.644625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0055-01</th>\n",
       "      <td>-0.124925</td>\n",
       "      <td>-0.197893</td>\n",
       "      <td>-0.132694</td>\n",
       "      <td>0.704438</td>\n",
       "      <td>0.154763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2483-01</th>\n",
       "      <td>-0.133543</td>\n",
       "      <td>-0.174587</td>\n",
       "      <td>-0.103291</td>\n",
       "      <td>1.473420</td>\n",
       "      <td>0.669303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2485-01</th>\n",
       "      <td>-0.147052</td>\n",
       "      <td>-0.072888</td>\n",
       "      <td>-0.213119</td>\n",
       "      <td>4.405612</td>\n",
       "      <td>11.503035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2486-01</th>\n",
       "      <td>-0.145321</td>\n",
       "      <td>-0.181076</td>\n",
       "      <td>-0.147395</td>\n",
       "      <td>1.013468</td>\n",
       "      <td>0.117745</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        1        10       100      1000      10000\n",
       "sample_id                                                         \n",
       "TCGA-02-0047-01 -0.144100 -0.136450 -0.207065  1.049402   0.644625\n",
       "TCGA-02-0055-01 -0.124925 -0.197893 -0.132694  0.704438   0.154763\n",
       "TCGA-02-2483-01 -0.133543 -0.174587 -0.103291  1.473420   0.669303\n",
       "TCGA-02-2485-01 -0.147052 -0.072888 -0.213119  4.405612  11.503035\n",
       "TCGA-02-2486-01 -0.145321 -0.181076 -0.147395  1.013468   0.117745"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(rnaseq_df.shape)\n",
    "rnaseq_df.iloc[:5, :5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_BARCODE</th>\n",
       "      <th>SAMPLE_BARCODE</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>SUBTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-OR-A5J1</td>\n",
       "      <td>TCGA-OR-A5J1-01</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Not_Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-OR-A5J2</td>\n",
       "      <td>TCGA-OR-A5J2-01</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Not_Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-OR-A5J3</td>\n",
       "      <td>TCGA-OR-A5J3-01</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Not_Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-OR-A5J5</td>\n",
       "      <td>TCGA-OR-A5J5-01</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Not_Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-OR-A5J6</td>\n",
       "      <td>TCGA-OR-A5J6-01</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Not_Applicable</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  PATIENT_BARCODE   SAMPLE_BARCODE DISEASE         SUBTYPE\n",
       "0    TCGA-OR-A5J1  TCGA-OR-A5J1-01     ACC  Not_Applicable\n",
       "1    TCGA-OR-A5J2  TCGA-OR-A5J2-01     ACC  Not_Applicable\n",
       "2    TCGA-OR-A5J3  TCGA-OR-A5J3-01     ACC  Not_Applicable\n",
       "3    TCGA-OR-A5J5  TCGA-OR-A5J5-01     ACC  Not_Applicable\n",
       "4    TCGA-OR-A5J6  TCGA-OR-A5J6-01     ACC  Not_Applicable"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_freeze_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>5S_rRNA</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPLE_BARCODE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0047-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0055-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2483-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2485-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2486-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 5S_rRNA  A1BG  A1CF  A2M  A2ML1\n",
       "SAMPLE_BARCODE                                  \n",
       "TCGA-02-0047-01        0     0     0    0      0\n",
       "TCGA-02-0055-01        0     0     0    0      0\n",
       "TCGA-02-2483-01        0     0     0    0      0\n",
       "TCGA-02-2485-01        0     0     0    0      0\n",
       "TCGA-02-2486-01        0     0     0    0      0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutation_df.iloc[:5, :5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9074, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>status</th>\n",
       "      <th>cancer_type</th>\n",
       "      <th>subtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPLE_BARCODE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0047-01</th>\n",
       "      <td>0</td>\n",
       "      <td>GBM</td>\n",
       "      <td>IDHwt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0055-01</th>\n",
       "      <td>1</td>\n",
       "      <td>GBM</td>\n",
       "      <td>IDHwt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2483-01</th>\n",
       "      <td>1</td>\n",
       "      <td>GBM</td>\n",
       "      <td>IDHmut-non-codel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2485-01</th>\n",
       "      <td>1</td>\n",
       "      <td>GBM</td>\n",
       "      <td>IDHwt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2486-01</th>\n",
       "      <td>0</td>\n",
       "      <td>GBM</td>\n",
       "      <td>IDHwt</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 status cancer_type           subtype\n",
       "SAMPLE_BARCODE                                       \n",
       "TCGA-02-0047-01       0         GBM             IDHwt\n",
       "TCGA-02-0055-01       1         GBM             IDHwt\n",
       "TCGA-02-2483-01       1         GBM  IDHmut-non-codel\n",
       "TCGA-02-2485-01       1         GBM             IDHwt\n",
       "TCGA-02-2486-01       0         GBM             IDHwt"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_df = (mutation_df\n",
    "    .loc[:, [gene]]\n",
    "    .merge(sample_freeze_df, left_index=True, right_on='SAMPLE_BARCODE')\n",
    "    .drop(columns='PATIENT_BARCODE')\n",
    "    .set_index('SAMPLE_BARCODE')\n",
    "    .rename(columns={gene: 'status',\n",
    "                     'DISEASE': 'cancer_type',\n",
    "                     'SUBTYPE': 'subtype'})\n",
    ")\n",
    "print(y_df.shape)\n",
    "y_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9074, 16148)\n",
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>10</th>\n",
       "      <th>100</th>\n",
       "      <th>1000</th>\n",
       "      <th>10000</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPLE_BARCODE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0047-01</th>\n",
       "      <td>-0.144100</td>\n",
       "      <td>-0.136450</td>\n",
       "      <td>-0.207065</td>\n",
       "      <td>1.049402</td>\n",
       "      <td>0.644625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0055-01</th>\n",
       "      <td>-0.124925</td>\n",
       "      <td>-0.197893</td>\n",
       "      <td>-0.132694</td>\n",
       "      <td>0.704438</td>\n",
       "      <td>0.154763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2483-01</th>\n",
       "      <td>-0.133543</td>\n",
       "      <td>-0.174587</td>\n",
       "      <td>-0.103291</td>\n",
       "      <td>1.473420</td>\n",
       "      <td>0.669303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2485-01</th>\n",
       "      <td>-0.147052</td>\n",
       "      <td>-0.072888</td>\n",
       "      <td>-0.213119</td>\n",
       "      <td>4.405612</td>\n",
       "      <td>11.503035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2486-01</th>\n",
       "      <td>-0.145321</td>\n",
       "      <td>-0.181076</td>\n",
       "      <td>-0.147395</td>\n",
       "      <td>1.013468</td>\n",
       "      <td>0.117745</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        1        10       100      1000      10000\n",
       "SAMPLE_BARCODE                                                    \n",
       "TCGA-02-0047-01 -0.144100 -0.136450 -0.207065  1.049402   0.644625\n",
       "TCGA-02-0055-01 -0.124925 -0.197893 -0.132694  0.704438   0.154763\n",
       "TCGA-02-2483-01 -0.133543 -0.174587 -0.103291  1.473420   0.669303\n",
       "TCGA-02-2485-01 -0.147052 -0.072888 -0.213119  4.405612  11.503035\n",
       "TCGA-02-2486-01 -0.145321 -0.181076 -0.147395  1.013468   0.117745"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_df = rnaseq_df.reindex(y_df.index)\n",
    "print(X_df.shape)\n",
    "print(X_df.isna().sum().sum())\n",
    "X_df.iloc[:5, :5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subset genes by mean absolute deviation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>126695</td>\n",
       "      <td>0.821090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>147798</td>\n",
       "      <td>0.808853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10053</td>\n",
       "      <td>0.808764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>54845</td>\n",
       "      <td>0.803626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>582</td>\n",
       "      <td>0.802497</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index         0\n",
       "0  126695  0.821090\n",
       "1  147798  0.808853\n",
       "2   10053  0.808764\n",
       "3   54845  0.803626\n",
       "4     582  0.802497"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first subset by MAD\n",
    "mad_genes_df = (rnaseq_df\n",
    "    .mad(axis=0)\n",
    "    .sort_values(ascending=False)\n",
    "    .reset_index()\n",
    ")\n",
    "mad_genes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['126695' '147798' '10053' '54845' '582']\n"
     ]
    }
   ],
   "source": [
    "mad_genes_df.columns=['gene_id', 'mad']\n",
    "mad_genes = mad_genes_df.iloc[:mad_threshold, :].gene_id.astype(str).values\n",
    "print(mad_genes[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9074, 100)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>126695</th>\n",
       "      <th>147798</th>\n",
       "      <th>10053</th>\n",
       "      <th>54845</th>\n",
       "      <th>582</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPLE_BARCODE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0047-01</th>\n",
       "      <td>-1.307892</td>\n",
       "      <td>-0.977865</td>\n",
       "      <td>-1.119059</td>\n",
       "      <td>-1.116767</td>\n",
       "      <td>0.458711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0055-01</th>\n",
       "      <td>-1.305820</td>\n",
       "      <td>-0.964695</td>\n",
       "      <td>-1.111900</td>\n",
       "      <td>-1.114346</td>\n",
       "      <td>-0.551517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2483-01</th>\n",
       "      <td>-1.307892</td>\n",
       "      <td>-0.984642</td>\n",
       "      <td>-1.119059</td>\n",
       "      <td>-1.116851</td>\n",
       "      <td>-0.470321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2485-01</th>\n",
       "      <td>-1.306134</td>\n",
       "      <td>-0.984678</td>\n",
       "      <td>-1.118344</td>\n",
       "      <td>-1.113535</td>\n",
       "      <td>1.100724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2486-01</th>\n",
       "      <td>-1.305639</td>\n",
       "      <td>-0.983574</td>\n",
       "      <td>-1.119059</td>\n",
       "      <td>-1.117114</td>\n",
       "      <td>1.006310</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   126695    147798     10053     54845       582\n",
       "SAMPLE_BARCODE                                                   \n",
       "TCGA-02-0047-01 -1.307892 -0.977865 -1.119059 -1.116767  0.458711\n",
       "TCGA-02-0055-01 -1.305820 -0.964695 -1.111900 -1.114346 -0.551517\n",
       "TCGA-02-2483-01 -1.307892 -0.984642 -1.119059 -1.116851 -0.470321\n",
       "TCGA-02-2485-01 -1.306134 -0.984678 -1.118344 -1.113535  1.100724\n",
       "TCGA-02-2486-01 -1.305639 -0.983574 -1.119059 -1.117114  1.006310"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_df = X_df.reindex(mad_genes, axis='columns')\n",
    "print(X_df.shape)\n",
    "X_df.iloc[:5, :5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate pan-cancer univariate feature correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2.76606990e+02 7.02872243e+01 3.47539437e+01 7.53519282e+02\n",
      " 7.44360880e+02 4.46192872e+02 1.17054890e+03 1.68578528e+02\n",
      " 5.27937846e+00 1.50456916e+02 1.69441254e+01 2.83799667e+01\n",
      " 1.81283292e+02 1.81860047e+02 1.21528596e+03 9.66673163e+01\n",
      " 9.73105305e+00 3.64416541e+02 2.31692409e+01 4.81917143e+02\n",
      " 6.16061787e+02 8.16934252e+01 1.57162530e+02 1.06114720e+03\n",
      " 9.55631077e+01 5.46751089e+02 7.44611032e+02 5.66550812e+02\n",
      " 5.52401563e+02 2.87701791e+00 2.82099162e+02 1.92113447e+02\n",
      " 1.49318434e+02 2.46996711e+01 1.57171729e+01 5.72854481e+00\n",
      " 9.59491476e+01 1.17693128e+01 1.43706915e+01 1.16411845e+02\n",
      " 2.24732308e+02 3.02421764e+02 4.40544808e+02 6.95960404e+00\n",
      " 4.69583295e+02 3.21183714e+01 6.30831691e+01 1.32760829e+02\n",
      " 7.49973212e+02 2.34246195e+00 4.05078465e+02 9.23233478e+02\n",
      " 1.90317914e+01 8.62842986e+02 1.29750470e+02 1.12404161e+03\n",
      " 3.81042096e+02 1.56625295e+02 1.35075994e+01 8.21075427e+02\n",
      " 6.59651258e+02 5.93133112e+02 7.64606987e+02 1.48822318e+02\n",
      " 2.11273531e+01 2.07376738e+02 8.07072450e+02 1.81911505e+02\n",
      " 3.98405466e+02 7.36515306e+02 5.49475226e+02 1.35902008e+01\n",
      " 3.48984479e+02 7.47111215e-03 7.89777604e+02 1.69890266e+02\n",
      " 1.77163113e+02 1.68689080e+01 8.23754151e+02 6.11847961e+02\n",
      " 1.31798792e+00 1.04125514e+02 5.36416470e+01 4.69419712e+01\n",
      " 8.25648964e+01 1.03334190e+02 4.45360156e+02 5.46030237e+02\n",
      " 1.43300446e+02 4.44156571e-01 4.82224299e+00 1.41563917e+02\n",
      " 4.12992084e+02 5.14887214e+01 1.47930403e+02 1.95332463e+01\n",
      " 2.39525957e+02 6.13166338e+00 1.66002798e+02 6.48574417e+01]\n"
     ]
    }
   ],
   "source": [
    "# now get univariate feature correlations with labels\n",
    "from sklearn.feature_selection import f_classif\n",
    "\n",
    "f_stats_pancan = f_classif(X_df, y_df.status)[0]\n",
    "print(f_stats_pancan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(33, 100)\n",
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>126695</th>\n",
       "      <th>147798</th>\n",
       "      <th>10053</th>\n",
       "      <th>54845</th>\n",
       "      <th>582</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KICH</th>\n",
       "      <td>0.481428</td>\n",
       "      <td>0.417635</td>\n",
       "      <td>0.012031</td>\n",
       "      <td>1.429834</td>\n",
       "      <td>0.103799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UCS</th>\n",
       "      <td>0.416370</td>\n",
       "      <td>0.000413</td>\n",
       "      <td>0.005767</td>\n",
       "      <td>0.065855</td>\n",
       "      <td>3.112278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACC</th>\n",
       "      <td>0.444091</td>\n",
       "      <td>0.914762</td>\n",
       "      <td>0.002503</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>0.218138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCPG</th>\n",
       "      <td>0.025548</td>\n",
       "      <td>0.026923</td>\n",
       "      <td>0.006727</td>\n",
       "      <td>0.008308</td>\n",
       "      <td>0.014604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pancan</th>\n",
       "      <td>276.606990</td>\n",
       "      <td>70.287224</td>\n",
       "      <td>34.753944</td>\n",
       "      <td>753.519282</td>\n",
       "      <td>744.360880</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            126695     147798      10053       54845         582\n",
       "KICH      0.481428   0.417635   0.012031    1.429834    0.103799\n",
       "UCS       0.416370   0.000413   0.005767    0.065855    3.112278\n",
       "ACC       0.444091   0.914762   0.002503    0.000216    0.218138\n",
       "PCPG      0.025548   0.026923   0.006727    0.008308    0.014604\n",
       "pancan  276.606990  70.287224  34.753944  753.519282  744.360880"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_f_stats_for_cancer_types(X_df, y_df):\n",
    "    f_stats_df = {}\n",
    "    for cancer_type in y_df.cancer_type.unique():\n",
    "        ct_samples = y_df[y_df.cancer_type == cancer_type].index\n",
    "        X_ct_df = X_df.reindex(ct_samples)\n",
    "        y_ct_df = y_df.reindex(ct_samples)\n",
    "        \n",
    "        with warnings.catch_warnings():\n",
    "            warnings.filterwarnings('error')\n",
    "            try:\n",
    "                f_stats_df[cancer_type] = f_classif(X_ct_df, y_ct_df.status)[0]\n",
    "            except RuntimeWarning:\n",
    "                # this can happen if there are no mutated samples in the cancer type\n",
    "                # in that case, just skip it\n",
    "                continue\n",
    "        \n",
    "    return pd.DataFrame(f_stats_df, index=X_df.columns).T\n",
    "\n",
    "f_stats_df = get_f_stats_for_cancer_types(X_df, y_df)\n",
    "f_stats_df.loc['pancan'] = f_stats_pancan\n",
    "print(f_stats_df.shape)\n",
    "print(f_stats_df.isna().sum().sum())\n",
    "f_stats_df.iloc[-5:, :5]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:pancancer-evaluation]",
   "language": "python",
   "name": "conda-env-pancancer-evaluation-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
